packages/flutter_tools/lib/src/device.dart

```diff
@@ -915,6 +915,42 @@ class DebuggingOptions {
 
   bool get hasObservatoryPort => hostVmServicePort != null;
 
+  List<String> getIOSLaunchArguments(EnvironmentType environmentType, String? route,  Map<String, Object?> platformArgs) {
+    final String dartVmFlags = computeDartVmFlags(this);
+    return <String>[
+      '--enable-dart-profiling',
+      if (environmentType == EnvironmentType.physical) '--disable-service-auth-codes',
+      if (environmentType == EnvironmentType.simulator && disableServiceAuthCodes) '--disable-service-auth-codes',
+      if (environmentType == EnvironmentType.physical && disablePortPublication) '--disable-observatory-publication',
+      if (startPaused) '--start-paused',
+      if (dartVmFlags.isNotEmpty) '--dart-flags=$dartVmFlags',
+      if (useTestFonts) '--use-test-fonts',
+      if (environmentType == EnvironmentType.physical && debuggingEnabled) ...<String>[
+        '--enable-checked-mode',
+        '--verify-entry-points',
+      ],
+      if (environmentType == EnvironmentType.simulator && buildInfo.isDebug) ...<String>[
+        '--enable-checked-mode',
+        '--verify-entry-points',
+      ],
+      if (enableSoftwareRendering) '--enable-software-rendering',
+      if (traceSystrace) '--trace-systrace',
+      if (skiaDeterministicRendering) '--skia-deterministic-rendering',
+      if (traceSkia) '--trace-skia',
+      if (traceAllowlist != null) '--trace-allowlist="$traceAllowlist"',
+      if (traceSkiaAllowlist != null) '--trace-skia-allowlist="$traceSkiaAllowlist"',
+      if (endlessTraceBuffer) '--endless-trace-buffer',
+      if (dumpSkpOnShaderCompilation) '--dump-skp-on-shader-compilation',
+      if (verboseSystemLogs) '--verbose-logging',
+      if (cacheSkSL) '--cache-sksl',
+      if (purgePersistentCache) '--purge-persistent-cache',
+      if (route != null) '--route=$route',
+      if (environmentType == EnvironmentType.physical && (platformArgs['trace-startup'] as bool? ?? false)) '--trace-startup',
+      if (enableImpeller) '--enable-impeller',
+      if (environmentType == EnvironmentType.simulator) '--observatory-port=${hostVmServicePort ?? 0}',
```

Comment:
I think it can go back to the behavior in https://github.com/flutter/flutter/pull/49735/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52L334 and pass the port along, if it's been set:
```suggestion
      if (hasObservatoryPort) '--observatory-port=${hostVmServicePort}',
```

Reply:
Should this differ between simulator and physical?

Looking at that change, looks like they used `deviceVmServicePort` https://github.com/flutter/flutter/pull/49735/files#diff-5a3d215e39f292ac706d495f042ef45fab9f13962022f4e8e4580254e8557d52L335

Whereas it's `hostVmServicePort` in original simulator args https://github.com/flutter/flutter/blob/01507bab565056b3dee933e2ff146966e4e8b001/packages/flutter_tools/lib/src/ios/simulators.dart#L465.

Is it possible to have both a `deviceVmServicePort` and `hostVmServicePort` and if so, which should be used if both are provided?

Reply:
Good question, I had to ask @jonahwilliams for clarification here, this behavior has changed over time and not every spot it's used has been kept consistent.

To quote Jonah:
> * Application runs on device, it starts up a VM Service instance. This instance will broadcast on a port + auth code combo.
> * (Optional) If Host/Device are different machines, we forward the device port to a selected host port.
> * (Optional) If dds is enabled, dds connects to the selected host port, then broadcasts on the dds port

For physical devices, the right thing to use is `deviceVmServicePort`.  The confusion is what to do if the "device" and host are really the same, as is the case with desktop apps or the iOS simulator, because there's no need to do  port forwarding if they are the same machines.  So I _think_ `hostVmServicePort` is correct in this case, though it's very confusing and it's likely we have some bugs in here.  

@christopherfujino the host/device port logic could use cleanup, comments, better help text, perhaps variable renaming, and verification we're doing the right thing for all device types.

Reply:
> @christopherfujino the host/device port logic could use cleanup, comments, better help text, perhaps variable renaming, and verification we're doing the right thing for all device types.

Agreed, let me make a tracking issue.

Reply:
> Agreed, let me make a tracking issue.

https://github.com/flutter/flutter/issues/114845